Homework2: Maksim Kvach

nkvamaks opened the pull request:
### PR checklist
Please make sure you've passed PR checklist
- [x] my PR passes CI checks (flake8 and pycodestyle)
- [x] my PR named according to guidelines.
- [x] my PR contains complete tasks descriptions for all tasks
- [x] my PR made from the fork feature branch to maxim homework# branch.
- [x] my PR contains only code that solves appropriate homework tasks. Redundant 
files (configs, etc. ) are put to .gitignore

nkvamaks committed:
Few time to solve more tasks

nkvamaks committed:
pydocstyle is checked

nkvamaks committed:
Edited text width.

nkvamaks committed:
Add comments in task1.py

nkvamaks committed:
Optimize code and add some comments in task2.py

nkvamaks committed:
Add comments in task3.py

nkvamaks committed:
Make readable code and add comments in task4.py

nkvamaks committed:
Slightly change comments in task5.py

nkvamaks committed:
Add comments in task6.py

nkvamaks committed:
Add Exercise 1 into a task7.py

nkvamaks committed:
Add Exercise 2 into a task7.py

nkvamaks committed:
Merge branch 'homework2' of https://github.com/belovmd/it-academy-summer into hw2_exp

nkvamaks committed:
Add 'empty string' check in task2.py

nkvamaks committed:
Merge branch 'homework2' of https://github.com/belovmd/it-academy-summer into hw2

nkvamaks committed:
Fix task2.py

nkvamaks committed:
Add Selector of Exercises in task7.py

There should be maximum five exercises inside task7.py. Selector
executes exercise with a number entered from console.

nkvamaks committed:
Add Exercise 3 (Valid Parentheses) in task7.py

nkvamaks committed:
Add Exercise 4 (HEX to RGB) in task7.py

belovmd commented on src/homework2/task2.py:
```diff
@@ -10,14 +12,25 @@ def longest_word(str_):
     # write your code here
-    return ''  # write return value here
+    str_1 = ''
+    for char in str_:  # убираем все не буквы и не пробелы
```
Appolo13 wouldn't pass it

belovmd replied:
we shouldn't ignore numbers

nkvamaks replied:
I was in doubt while creating first version of the function whether I should ignore numbers or not because a) I couldn't imagine at a glance words together with numbers and b) counting numbers wasn't explicitly indicated in the task description. Now I see that there are some words including numbers like 'b2b', '4U', 'U2', hashes etc. But for the record, 'Apollo 13' is usually written as two distinct words.

belovmd commented on src/homework2/task2.py:
```diff
@@ -1,6 +1,8 @@
-"""Найти самое длинное слово в введенном предложении. В случае если их
-    несколько, самое левое в строке Учтите что в предложении есть знаки
-    препинания.
+"""
```
overall solution looks too complex

nkvamaks replied:
Drastically simplified.

belovmd commented on src/homework2/task5.py:
```diff
@@ -9,12 +12,22 @@ def fibonacci(n):
+    a_0 = 1  # первое значение последовательности
+    a_1 = 1  # второе значение последовательности
+
+    if n == 1:  # частный случай
```
```
if n in (1, 2):
  return 1
```
much more simple?

nkvamaks replied:
Update the task with a simplified code.

belovmd commented on src/homework2/task7.py:
```diff
@@ -0,0 +1,173 @@
+    not contain any parentheses at all. Do not treat other forms of brackets
+    as parentheses (e.g. [], {}, <>).
+    """
+    print("Exercise 3: Valid Parentheses.")
```
it looks pretty bulky.  mostly because of comments. 
overall good solution.

nkvamaks replied:
Cheers!

belovmd commented on src/homework2/task7.py:
```diff
@@ -0,0 +1,173 @@
+    print("Exercise 4: Convert A Hex String To RGB.")
+    str_ = input("Enter a hexadecimal color string with a following format"
+                 " (e.g. #FF0056): ")
+    if re.match(r'#[0-9a-fA-F]{6}', str_):
```
why do we need this validation? according to the task description, we receive valid string.

nkvamaks replied:
Basically the validation was done to mimic real case when input can be made with mistakes. But you're right, for synthetic task the validation is redundant. So it was removed from code.

belovmd reviewed (commented):
Thank you for your work. 
Please address next paragraphs from review notes [Review Notes](https://github.com/belovmd/it-academy-summer/blob/master/Homework%20review/Homework2.md):
- 6 all comments inside tasks looks redundant
- 11
- 18

nkvamaks committed:
Update tasks in homework2

task1.py - was written mainly as one string of code
task2.py - was drastically simplified using max() function
to find biggest word in the list. Now the function get rid of
just punctuation and leave numbers with text symbols intact.
task4.py - counting of upper- and lower-case symbols is made using 
'generator comprehensions' instead of 'if conditions'.
task5.py - was simplified a bit
task7.py - validation in Exercise 4 was removed.

nkvamaks committed:
Fix 'empty string' case in task2.py

belovmd commented on src/homework2/task2.py:
```diff
@@ -10,14 +12,16 @@ def longest_word(str_):
     # write your code here
-    return ''  # write return value here
+    import string
+    # get rid of punctuation in the string and create a list of single words
```
redundant comments

belovmd commented on src/homework2/task3.py:
```diff
@@ -12,12 +13,15 @@ def sub_string(str_):
     # write your code here
-    return ''  # write return value here
+    str_1 = ''
+    for char in str_:  # create a new string of unique chars w/o spaces
```
comments

belovmd commented on src/homework2/task7.py:
```diff
@@ -0,0 +1,168 @@
+    print("Exercise 3: Valid Parentheses.")
+    str_ = input("Enter a String with Parentheses '()': ")
+    parenth = '()'  # for entry check
+    stack = []  # stack where parentheses will be saved
```
comments

belovmd reviewed (commented):
8,5
not addressed review tasks. 

Overall it's a good work
